source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -237,7 +241,16 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_st
 
   FilterHeadersStatus status = FilterHeadersStatus::Continue;
   if (decoding_state_.sendHeaders()) {
-    status = onHeaders(decoding_state_, headers, end_stream);
+    ProtobufWkt::Struct proto;
+
+    if (config_->expressionManager().hasRequestExpr()) {
+      auto activation_ptr = Filters::Common::Expr::createActivation(decoding_state_.streamInfo(),
+                                                                    &headers, nullptr, nullptr);
+      proto = config_->expressionManager().evaluateRequestAttributes(*activation_ptr);
+    }
+
+    status = onHeaders(decoding_state_, headers, end_stream,
+                       config_->expressionManager().hasRequestExpr() ? &proto : nullptr);
```

Comment:
This might be a bit cleaner, but it doubles the number of places `onHeaders` may be called, which can add mental overhead when debugging and/or reading the code. WDYT @tyxia?
```suggestion
    if (config_->expressionManager().hasRequestExpr()) {
      auto activation_ptr = Filters::Common::Expr::createActivation(decoding_state_.streamInfo(),
                                                                    &headers, nullptr, nullptr);
      auto proto = config_->expressionManager().evaluateRequestAttributes(*activation_ptr);
      status = onHeaders(decoding_state_, headers, end_stream, &proto);
    } else {
      status = onHeaders(decoding_state_, headers, end_stream, nullptr);
    }
```

Reply:
TBH, either works for me. I think the code you currently have LGTM.